### PR TITLE
New overlapping criteria, fluffiness, and prm cleanup

### LINKED ISSUE #84: `LAYER_BASE_FRAC` not fed to `self.metarize()`
**Describe the bug**
This bug was reported by @loforest in #83. The parameter `LAYER_BASE_FRAC` is never fed to the `self.metarize()` routine of the slicing, grouping, and layering steps.

### LINKED ISSUE #85: `test_scientific_stability()` should always process all cases, even if one fails
**Describe the change**
At the moment, `test-scientific_stability()` checks all cases sequentially, and stops as soon as one of them fails. 

This is not ideal, as fixing one may then break another, etc ... 

A better approach would be to create some sort of test factory, such that each case is assessed individually independently of the others.

### LINKED ISSUE #87: Pandas v1.5.0 FutureWarning
**Describe the change**
The Github CI/CD checks are picking-up a new `FutureWarning` from pandas coming from v1.5.0. 

I will fix those warnings using backward compatible code (hopefully). 

To be safe, I could bump the minimum required version of pandas to be v1.5.0. @nestabur, in a general sense, how problematic/annoying is it on the autometpy side of things if ampycloud has new package version requirements ?

### LINKED ISSUE #89: Compute the fluffiness of cloud layers
**Describe the change**

Certain sets of ceilometer hits can display clear structure, possibly with ordered altitude fluctuations. Alternatively, ceilometer hits can be spread throughout the entire spatial extent of a given slice/group/layer. 

To aid in differentiating between these two types of structures (and possibly pave the way for better algorithms), we bring forth the idea of **"fluffiness"**, ~~ranging between 0 (perfectly structured) to 1 (ultra-fluffy). This number is computed as the ratio between:~~
  ~~1) the area of the complex hull encompassing all the ceilometer hits in the slice/group/layer, and~~
  ~~2) the full extent (in ft x s space) of the ceilometer hits span, i.e. (alt_max - alt_min) * (dt_max - dt_min).~~

**Edit:** the initial idea of the Convex Hull was abandoned in favor of LOWESS. See comments below for details.

### LINKED ISSUE #90: `step_scale` crashes with a single scale
**Describe the bug**
Specifying a single scale (and thus no steps) crashes the `step_scale()` routine.

### LINKED ISSUE #92: Specify `OKTA_LIM0/8` in absolute hit counts rather than percentages
**Describe the change**
@loforest asked to have the `OKTA_LIM0` and `OKTA_LIM8` parameters in absolute counts rather than percentages.

### LINKED ISSUE #93: Use new Issue forms
**Describe the change**
The new [issue forms](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository#creating-issue-forms) would be a nice improvement to the current issue and feature request templates.

### LINKED ISSUE #95: Adapt pylinter CI/CD action for pylint 2.16
**Describe the bug**
The `pylinter.py` script crashes following a new DeprecationWarning in pylint 2.16+. 

This needs to be fixed. See [this dvas commit](https://github.com/MeteoSwiss/dvas/pull/274/commits/20c8e1d29519d987f79b0390a7bf549fe90d6b19) for the solution.

### LINKED ISSUE #98: pkg_resources is deprecated as an API
**Describe the bug**
The ``test_version.py'' test relies on pkg_resources, which is deprecated. This triggers warnings when running the ampycloud tests.

**Expected behavior**
No warnings.

### LINKED ISSUE #99: Trigger the pytest workflow at regular (automated) intervals
**Describe the change**
At the moment, the ampycloud actions (and in particular the one testing the code) are only run when a pull request is being issued on Github.

It would however be beneficial to have some of these actions executed more regularly to anticipate any issue as the code dependencies evolve.

One (easy) option is to trigger the workflow at [regular intervals](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule) automatically (e.g. midnight every Monday ?). The results of the latest execution could then be grabbed remotely when needed.

However, it is also possible to [interact directly with Github Actions via REST API](https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow#running-a-workflow-using-the-rest-api). @regDaniel  @loforest @nestabur : perhaps this could be the better option, to let you/autometpy decide when to trigger the ampycloud tests to assess if they still all pass with current dependency versions ?

### LINKED ISSUE #100: Add support for pandas v2.x
**Describe the bug**
The code needs a cleanup following the arrival of pandas v2.x.

### LINKED ISSUE #102: Format demo dataset as ASCII instead of ``.pkl``
**Describe the change**
The update of pandas to v2.x makes it impossible to load the ``.pkl`` file containing the demo/test datasets for ampycloud. 

This is a terrible thing. The test data should be stored as readable ASCII files and converted by the test suite instead, to avoid similar problems in the future.

### PR DESCRIPTION
**Pull request by fpavogt**
**Description:**

This PR introduces a series of updates to the code, following the suggestions by @loforest in #83, including:

- the default value for `OKTA_LIM8` has been changed to `99` following the large scale tests of @loforest.
- the GROUPING affinity has been changed from `manhattan` to `euclidean`, since the former is detrimental to rising/dropping layers.
- the "vertical extrema" of slices (to assess if they overlap) are now computed as `min/max +- alt_pad_perc * (thickness)`. **`alt_pad_perc` is a new ampycloud `GROUPING` parameter that replaces `overlap` !**
- the minimum-required separation between layers can now be adjusted as a function of altitude, and is no longer hardcoded to +- 10'000 ft.
- the cloud base is now computed as a percentile of the total amount of points (starting from the smallest ones), with the option to define a lookback percentile to only use a fraction of the most recent measurements.
- the `metarize` routine now also computes the "fluffiness" of layers (see #89). This requires to add `scipy`  and `statsmodels` to the list of required dependencies.
- the list of user-tunable parameters has been cleaned from all the superfluous ones.
- the parameter `OKTA_LIM0` has been replaced by `MAX_HITS_OKTA0`, which now specifies the maximum (absolute) number of hits in a given slice/group/layer for it to be considered to be 0 oktas.
- the parameter `OKTA_LIM8` has been replaced by `MAX_HOLES_OKTA8`, which now specifies the maximum (absolute) number of holes (= non-detection of cloud bases) in a given slice/group/layer for it to be considered to be 8 oktas.
- layers are now required to be 250ft apart (below 10'000 ft) in order for a group to be split.

It also contains other less critical changes, including:
- improved docstrings
- more robust tests
- bug fix in `step_scale()`

**Error(s) fixed:**

Fixes #85 (scientific validity tests are now always run fully, even if one fails). Fixes #84. Fixes #89. Fixes #90. Fixes #87. Fixes #92. Fixes #98.

With the merging of #101 this also: fixes #102, fixes #95, fixes #100, fixes #93, and fixes #99.

**Checklists**:
- [x] New code includes dedicated tests.
- [x] New code has been linted.
- [x] New code follows the project's style.
- [x] New code is compatible with the 3-Clause BSD license.
- [x] CHANGELOG has been updated.
- [ ] AUTHORS has been updated.
- [x] Copyright years in module docstrings have been updated.

### CODE REVIEW

**Comment by fpavogt** [comment]
@loforest: ready for your review whenever you have time. You could also consider running the large scale tests directly via a fork of my branch, if you wanted to test the full impact of these changes before approving them.

**Comment by fpavogt** [comment]
Hey @loforest and @pjuda, 

The request in #92 is now implemented, with two re-named parameters `MAX_HITS_OKTA0` (default: 3) and `MAX_HOLES_OKTA8` (default: 1), that allow to set maximum number of hits/holes **in absolute terms** for the NCD and OVC categories (0 and 8 oktas).

This **did** have some effect on one of the scientific test, with the addition of a `FEW066` layer with 4 hits, that was ignored before:

Geneva_2021 10 06-19 20 00_FEW066-SCT073_MSA10000

If we decrease the default value of `MAX_HITS_OKTA0` to 2 (as discussed yesterday), then another scientific test will be affected (the bottom-most layer would not longer be ignored):

Geneva_2021 09 19-11 50 00_OVC032

I can easily set the default values to whatever you think is best. Maybe the large scale tests will provide some guidance on that front.

Long story short: the code in the `develop_vof` branch is ready to be "large-scale evaluated" whenever you want. The sphinx docs has been updated accordingly within the branch - it has not yet been released on the web of course.

I will also update the draft of the scientific article, but that has no impact on the code, of course.

Do not hesitate to ping me if anything is unclear/surprising when running the tests.

**Review by pjuda — COMMENTED** [review]
Hi @fpavogt ! That is a significant piece of work and very well done. Thank you! I have left some comments which might help to improve how the code reads. At this point, I cannot say much about the core logic and algorithms behind, would need more time to deep dive. Depends on @loforest how we move on with the review.

The "large scale evaluation" is on our list to do. Should we do it first and then come back to the review?

**Comment by fpavogt** [comment]
@pjuda: Thanks for the suggestions ! None of them will affect the behavior of the code, so I would suggest to go ahead with the large scale evaluation. I can then implement the suggested updates together with whatever may be warranted following the large scale tests.

File: test/ampycloud/test_fluffer.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    assert out[0] == 0
+    assert np.array_equal(out[1][:, 1], pts[:, 1])
+
+    # Test the LOWESS warning issued with duplicat points
+    pts = np.array([[0, 0], [0, 1], [1, 0]])
+    # For some reason, pytest refuses to convert the LOWESS warning into an error ...
+    # ... so I need to catch it a bit differently ... sigh !
+    with warnings.catch_warnings(record=True) as w_list:
+        # warnings.simplefilter("error")
```
[the remark this answers is not in the document]
Comment (fpavogt) [reply test/ampycloud/test_fluffer.py]: Done.

File: src/ampycloud/data.py [hunk 2]
```diff
@@ -304,13 +315,20 @@ def metarize(self, which: int = 'slices', base_frac: float = 0.1,
 
         """
 
+        # TODO: base_lvl_alt_perc, base_lvl_lookback_perc, lim0, lim8 are all accessible via
```
[the remark this answers is not in the document]
Comment (fpavogt) [reply src/ampycloud/data.py]: TODO replaced with a Note-to-self, as this is a general comment about the structure of the function.

File: src/ampycloud/fluffer.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    # Begin by sorting points if they are not already - I'll need this to compute the fluffiness
+    pts = pts[pts[:, 0].argsort()]
+
+    # Points with duplicated x values trigger a RuntimeWarning from LOWESS.
+    # To avoid it, let's jiggle duplicated points a little bit.
+    if len(np.unique(pts[:, 0])) == len(pts):
+        unique_xs = pts[:, 0]
+    else:
+        diffs = np.diff(pts[:, 0])
+        unique_xs = [pts[0, 0]] + [item if diffs[ind] > 0 else item + (ind + 1) * 1e-5
```
[the remark this answers is not in the document]
Comment (fpavogt) [reply src/ampycloud/fluffer.py:63]: Comment added: 
```
# Let's add (ind * 1e-5) to any duplicated point. We use the point index "ind"
# to make sure tri-/quadru-/multi-ple points all get shifted to distinct
# positions.
```

File: src/ampycloud/layer.py [hunk 4]
```diff
@@ -264,7 +264,7 @@ def ncomp_from_gmm(vals: np.ndarray,
         best_ncomp -= 1
 
     if not len(np.unique(best_ids)) == best_ncomp:
-        raise AmpycloudError('Ouch ! This error is impossible !')
+        raise AmpycloudError('This error is impossible !')
```
[the remark this answers is not in the document]
Comment (fpavogt) [reply src/ampycloud/layer.py]: Error replaced with an assert.

File: src/ampycloud/fluffer.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    if pts.ndim != 2:
+        raise AmpycloudError('pts should be 2-dimensional.')
+
+    # If I was given a single point, things are easy
+    if len(pts) == 1:
+        return 0, pts
+
+    # Begin by sorting points if they are not already - I'll need this to compute the fluffiness
+    pts = pts[pts[:, 0].argsort()]
```
[the remark this answers is not in the document]
Comment (fpavogt) [reply src/ampycloud/fluffer.py:52]: No, `pts.sort(axis=0)` is not doing the same thing.

File: src/ampycloud/data.py [hunk 6]
```diff
@@ -258,22 +258,29 @@ def max_hits_per_layer(self) -> int:
...
                 Defaults to 'slices'.
-            base_frac (float, optional): amount of the lowest slice/group/layer elements,
-                expressed as a fraction (0<=base_frac<=1) of the total hit count, to consider when
-                deriving the slice/group/layer altitude (as a median). Defaults to 0.1, i.e. the
-                cloud base is the median of the lowest 10% of cloud hits of that slice/group/layer.
-            lim0 (int|float, optional): upper limit of the sky coverage percentage for the 0 okta
-                bin, in %. Defaults to 0.
-            lim8 (int|float, optional): lower limit of the sky coverage percentage for the 8 okta
-                bin, in %. Defaults to 100.
+            base_lvl_alt_perc (float, optional): amount of the lowest slice/group/layer elements,
```
[the remark this answers is not in the document]
Comment (fpavogt) [reply src/ampycloud/data.py:270]: See #94 .

File: src/ampycloud/data.py [hunk 7]
```diff
@@ -258,22 +258,29 @@ def max_hits_per_layer(self) -> int:
...
+                expressed in percent of the total hit count, to ignore when
+                deriving the slice/group/layer altitude (as the min value of the remaining points).
+                Defaults to 5, i.e. the cloud base is the min alt value ignoring the lowest 5% of
+                cloud hits of that slice/group/layer.
+            base_lvl_lookback_perc (float, optional): amount of the slice/group/layer elements to
```
[the remark this answers is not in the document]
Comment (fpavogt) [reply src/ampycloud/data.py:275]: See #94.

**Comment by fpavogt** [comment]
Hey @loforest, @regDaniel, @pjuda, @sibalm - fyi, this PR is now ready once again to have its scientific stability/validity evaluated. I have just merged #101 into it (which included a handful of small "logistical" Actions/Tests tweaks) so that all the tests are now green again. 

Happy to talk live about the stability tests when you want, and in particular the necessity (maybe?) to fine-tune some of the parameters (old and new).

Side-note: the pytest action will be triggered twice per week as soon as this gets merged into `develop`. Only then will its result be accessible via RestAPI.

File: test/ampycloud/test_core.py [hunk 8]
```diff
@@ -95,24 +95,24 @@ def test_run_single_point():
...
                         columns=['ceilo', 'dt', 'alt', 'type'])
     # Set the proper column types
     for (col, tpe) in hardcoded.REQ_DATA_COLS.items():
-        data.loc[:, col] = data.loc[:, col].astype(tpe)
+        data[col] = data.loc[:, col].astype(tpe)
 
     # Run the code
     out = run(data)
-    assert out.metar_msg() == 'SCT020'
+    assert out.metar_msg() == 'NCD'
```
Comment (pjuda) [inline test/ampycloud/test_core.py:109]: What would be a minimal test case with an output which is not NCD but actually some cloud layers identified?
